File: koauth-sample-scalatra/src/main/scala/com/hunorkovacs/koauthsamplescalatra/Twitter.scala
package com.hunorkovacs.koauthsamplescalatra

import _root_.akka.actor.ActorSystem
import com.hunorkovacs.koauth.domain.OauthParams._
import com.hunorkovacs.koauth.domain.{TokenResponse, KoauthRequest}
import com.hunorkovacs.koauth.service.Arithmetics._
import com.hunorkovacs.koauth.service.consumer.DefaultConsumerService
import org.scalatra._
import spray.client.pipelining
import spray.client.pipelining._
import spray.http.HttpHeaders.RawHeader
import spray.http._

import scala.concurrent.{Await, Future}
import scala.concurrent.duration._

class Twitter extends ScalatraServlet {

  private val ConsumerKey = "TZO4QQGaFJBjXuKsyh8n1KrpE"
  private val ConsumerSecret = "7t57qMb1unIyiZrRVuWwSwAQ6QXxcUtuRh98iqeHQSjfXtFsIx"
  private val RequestTokenUrl = "https://api.twitter.com/oauth/request_token"
  private val AuthorizeTokenUrl = "https://api.twitter.com/oauth/authorize"
  private val AccessTokenUrl = "https://api.twitter.com/oauth/access_token"

  implicit private val system = ActorSystem()
  import system.dispatcher
  private val pipeline: HttpRequest => Future[HttpResponse] = sendReceive
  private val consumer = new DefaultConsumerService(system.dispatcher)
  private var token = TokenResponse("", "")

  get("/requestToken") {
    val requestTokenResponseF = consumer.createRequestTokenRequest(
        KoauthRequest("POST", RequestTokenUrl, None), ConsumerKey, ConsumerSecret,
          "http://127.0.0.1:8080/accessToken") flatMap { requestWithInfo =>
      pipeline(pipelining.Post(RequestTokenUrl).withHeaders(RawHeader("Authorization", requestWithInfo.header)))
    } map { response =>
      System.out.println("Response: HTTP " + response.entity.asString)
      parseRequestTokenResponse(response.entity.asString).right.get
    }
    token = Await.result(requestTokenResponseF, 4 seconds)
    redirect(AuthorizeTokenUrl + "?" + TokenName + "=" + token.token)
  }

  get("/accessToken") {
    val accessTokenResponseF = consumer.createAccessTokenRequest(
      KoauthRequest("POST", AccessTokenUrl, None), ConsumerKey, ConsumerSecret, token.token, token.secret,
        params(VerifierName)) flatMap { requestWithInfo =>
      pipeline(pipelining.Post(AccessTokenUrl).withHeaders(RawHeader("Authorization", requestWithInfo.header)))
    } map { response =>
      System.out.println("Response: HTTP " + response.entity.asString)
      parseAccessTokenResponse(response.entity.asString).right.get
    }
    token = Await.result(accessTokenResponseF, 4 seconds)
    redirect("/lastTweet")
  }

  get("/lastTweet") {
    val lastTweetUrl = "https://api.twitter.com/1.1/statuses/user_timeline.json?count=1&include_rts=1&trim_user=true"
    val lastTweetResponseF = consumer.createOauthenticatedRequest(KoauthRequest("GET", lastTweetUrl, None),
        ConsumerKey, ConsumerSecret, token.token, token.secret) flatMap { requestWithInfo =>
      pipeline(pipelining.Get(lastTweetUrl).withHeaders(RawHeader("Authorization", requestWithInfo.header)))
    }
    Await.result(lastTweetResponseF, 4 seconds).entity.asString
  }
}

File: java-consumer-spark/src/main/java/com/hunorkovacs/koauth/sample/java/consumerspark/Twitter.java
package com.hunorkovacs.koauth.sample.java.consumerspark;

import com.hunorkovacs.koauth.domain.OauthParams;
import com.hunorkovacs.koauth.domain.TokenResponse;
import com.hunorkovacs.koauth.domain.KoauthRequest;
import com.hunorkovacs.koauthsync.service.consumer.ConsumerService;
import com.hunorkovacs.koauthsync.service.consumer.DefaultConsumerService;
import com.hunorkovacs.koauth.service.consumer.RequestWithInfo;
import com.hunorkovacs.koauth.service.Arithmetics;

import scala.Option;
import scala.concurrent.ExecutionContext;
import spark.Request;
import spark.Route;

import javax.ws.rs.client.*;
import javax.ws.rs.core.Response;

import static spark.Spark.*;

public class Twitter {

    private static final String CONSUMER_KEY = "E0XkGScbAci70CeBikSYh90EF";
    private static final String CONSUMER_SECRET = "3BIeeX56BuRng0Yq7MJCxaT15wH5YgJ1QmFkfGKw7yFPZbE3Eh";
    private static final String REQUEST_TOKEN_URL = "https://api.twitter.com/oauth/request_token";
    private static final String AUTHORIZE_TOKEN_URL = "https://api.twitter.com/oauth/authorize";
    private static final String ACCESS_TOKEN_URL = "https://api.twitter.com/oauth/access_token";

    private ExecutionContext ec = ExecutionContext.Implicits$.MODULE$.global();
    private Client http = ClientBuilder.newClient();
    private ConsumerService consumer = new DefaultConsumerService(ec);

    private TokenResponse requestTokenResponse;
    private TokenResponse accessTokenResponse;

    private class RequestTokenRoute implements Route {
        public Object handle(Request request, spark.Response response) throws Exception {
            Invocation.Builder builder = http.target(REQUEST_TOKEN_URL).request();
            RequestWithInfo requestWithInfo = consumer.createRequestTokenRequest(KoauthRequest.apply("POST",
                            REQUEST_TOKEN_URL, Option.<String>empty()),
                    CONSUMER_KEY,
                    CONSUMER_SECRET,
                    "http://127.0.0.1:4567/accessToken");
            Invocation invocation = builder.header("Authorization", requestWithInfo.header()).buildPost(Entity.text(""));

            Response twResponse = invocation.invoke();

            System.out.println("Response: HTTP " + twResponse.getStatus());
            String body = twResponse.readEntity(String.class);
            System.out.println(body);
            requestTokenResponse = Arithmetics.parseRequestTokenResponse(body).right().get();
            response.redirect(AUTHORIZE_TOKEN_URL +
                    "?" + OauthParams.TokenName() + "=" + requestTokenResponse.token());
            return response;
        }
    }

    private class AccessTokenRoute implements Route {
        public Object handle(Request request, spark.Response response) throws Exception {
            Invocation.Builder builder = http.target(ACCESS_TOKEN_URL).request();
            RequestWithInfo requestWithInfo = consumer.createAccessTokenRequest(KoauthRequest.apply("POST",
                            ACCESS_TOKEN_URL, Option.<String>empty()),
                    CONSUMER_KEY,
                    CONSUMER_SECRET,
                    requestTokenResponse.token(),
                    requestTokenResponse.secret(),
                    request.queryMap(OauthParams.VerifierName()).value());
            Invocation invocation = builder.header("Authorization", requestWithInfo.header()).buildPost(Entity.text(""));

            Response twResponse = invocation.invoke();

            System.out.println("Response: HTTP " + twResponse.getStatus());
            String body = twResponse.readEntity(String.class);
            System.out.println(body);
            accessTokenResponse =  Arithmetics.parseAccessTokenResponse(body).right().get();
            response.redirect("/lastTweet");
            return response;
        }
    }

    private class LastTweetRoute implements Route {
        public Object handle(Request request, spark.Response response) throws Exception {
            String lastTweetUrl = "https://api.twitter.com/1.1/statuses/user_timeline.json?count=1&include_rts=1&trim_user=true";
            Invocation.Builder builder = http.target(lastTweetUrl).request();
            RequestWithInfo requestWithInfo = consumer.createOauthenticatedRequest(KoauthRequest.apply("GET",
                            lastTweetUrl, Option.<String>empty()),
                    CONSUMER_KEY,
                    CONSUMER_SECRET,
                    accessTokenResponse.token(),
                    accessTokenResponse.secret());
            Invocation invocation = builder.header("Authorization", requestWithInfo.header()).buildGet();

            Response twResponse = invocation.invoke();

            System.out.println("Response: HTTP " + twResponse.getStatus());
            String body = twResponse.readEntity(String.class);
            System.out.println(body);
            return body;
        }
    }

    public static void main(String[] args) {
        Twitter twitter = new Twitter();
        get("/requestToken", twitter.new RequestTokenRoute());
        get("/accessToken", twitter.new AccessTokenRoute());
        get("/lastTweet", twitter.new LastTweetRoute());
    }
}
